torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,609 @@
+
+    const auto& val = [](const Tensor& t) {return t.item<double>(); };
+
+    auto d_norm = val(d.abs().max());
+    g = g.clone(at::MemoryFormat::Contiguous);
+    // evaluate objective and gradient using initial step
+    auto obj_func_res = obj_func(x, t, d);
+    auto f_new = std::get<0>(obj_func_res);
+    auto g_new = std::get<1>(obj_func_res);
+    int64_t ls_func_evals = 1;
+    auto gtd_new = g_new.dot(d);
+
+    // bracket an interval containing a point satisfying the Wolfe criteria
+    double t_prev = 0;
+    auto f_prev = f;
+    auto g_prev = g;
+    auto gtd_prev = gtd;
+    bool done = false;
+    auto ls_iter = 0;
+    std::vector<double> bracket, bracket_f;
+    std::vector<Tensor> bracket_g, bracket_gtd;
+
+    while (ls_iter < max_ls) {
+      // check conditions
+      if ((f_new > (f + c1 * t * val(gtd))) || (ls_iter > 1 && (f_new >= f_prev))) {
+        bracket = {t_prev, t};
+        bracket_f = {f_prev, f_new};
+        bracket_g = {g_prev, g_new.clone(at::MemoryFormat::Contiguous)};
+        bracket_gtd = {gtd_prev, gtd_new};
+        break;
+      }
+      if (abs(val(gtd_new)) <= (-c2 * val(gtd))) {
+        bracket = {t, t};
+        bracket_f = {f_new, f_new};
+        bracket_g = {g_new, g_new};
+        done = true;
+        break;
+      }
+      if (val(gtd_new) >= 0) {
+        bracket = {t_prev, t};
+        bracket_f = {f_prev, f_new};
+        bracket_g = {g_prev, g_new.clone(at::MemoryFormat::Contiguous)};
+        bracket_gtd = {gtd_prev, gtd_new};
+        break;
+      }
+      // interpolate
+      auto min_step = t + 0.01 * (t - t_prev); // NOLINT(cppcoreguidelines-avoid-magic-numbers)
+      auto max_step = t * 10; // NOLINT(cppcoreguidelines-avoid-magic-numbers)
+      auto tmp = t;
+      t = _cubic_interpolate(t_prev, f_prev, val(gtd_prev),
+                             t, f_new, val(gtd_new),
+                             std::make_tuple(min_step, max_step));
+      // next step
+      t_prev = tmp;
+      f_prev = f_new;
+      g_prev = g_new.clone(at::MemoryFormat::Contiguous);
+      gtd_prev = gtd_new;
+      obj_func_res = obj_func(x, t, d);
+      f_new = std::get<0>(obj_func_res);
+      g_new = std::get<1>(obj_func_res);
+      ls_func_evals += 1;
+      gtd_new = g_new.dot(d);
+      ls_iter += 1;
+    }
+    // reached max number of iterations?
+    if (ls_iter == max_ls) {
+      bracket = {0, t};
+      bracket_f = {f, f_new};
+      bracket_g = {g, g_new};
+    }
+
+    // zoom phase: we now have a point satisfying the criteria, or
+    // a bracket around it. We refine the bracket until we find the
+    // exact point satisfying the criteria
+    bool insuf_progress = false;
+    // find high and low points in bracket
+    int64_t low_pos, high_pos;
+    if (bracket_f[0] <= bracket_f[1]) {
+      low_pos = 0;
+      high_pos = 1;
+    } else {
+      low_pos = 1;
+      high_pos = 0;
+    }
```

Comment:
might be easier to write:
```suggestion
    std::tie(low_pos, high_pos) = bracket_f[0] <= bracket_f[1] ? std::make_tuple(0, 1) : std::make_tuple(1, 0);
```